### ISSUE TITLE
Class 'CRUD' not found during package:discover

### ISSUE THREAD

**Issue by pavoltanuska** [Laravel-Backpack/Settings#64]
# Bug report

### What I did:

I've tried to upgrade Backpack from 3.2.x to 3.3.x

### What I expected to happen:

Clean upgrade without any errors.

### What happened:

```
-> % composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 0 updates, 0 removals
Generating optimized autoload files
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover

In settings.php line 18:

  Class 'CRUD' not found

Script @php artisan package:discover handling the post-autoload-dump event returned with error code 1

```

This is probably caused by package autodiscovery - after forcing the changes back (next step) I've seen that the autodiscovery order is incorrect (settings are loaded before base and crud):

```
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover
...
Discovered Package: backpack/settings
...
Discovered Package: backpack/base
Discovered Package: backpack/crud
Package manifest generated successfully.
```

### What I've already tried to fix it:

- Reverted a change in `vendor/backpack/settings/src/routes/backpack/settings.php` (seen in https://github.com/Laravel-Backpack/Settings/commit/82905b0b3eb8bf8810d605c93fcc18bf40bbaafe#diff-07b8faea13131bb615e1a0b610347f0d) manually from `CRUD::resource(...)` back to `Route::resource(...)` - works ok
- set `$defer = true;` in `vendor/backpack/settings/src/SettingsServiceProvider.php` - works ok (and probably is the right solution)

### Backpack, Laravel, PHP, DB version:

- PHP 7.1.4
- laravel/framework v5.5.22
- backpack/base 0.8.2
- backpack/crud 3.3.1
- backpack/settings 2.0.25

**Comment by pavoltanuska** [comment]
oh, I take the suggested fixes back - after trying to run the admin interface in browser the issue came back. What I really did to work around it (this time for real I guess) was to put `backpack/settings` into dont-discover section in main `composer.json` file and load the `Backpack\Settings\SettingsServiceProvider::class` manually in `config/app.php`

**Comment by tabacitu** [comment]
Huh... this is odd. Just updated a project to test, then installed it on a fresh laravel - no such issue in my experience.

I wonder if the problem isn't the order, as defined in ```composer.json``` - is your settings requirement BEFORE the crud requirement?

Cheers!

**Comment by jguyomard** [comment]
Yes. I've this issue too, on some projects, because Laravel try to load backpack-setting before backpack-crud. According to me, you have to move this line in the `boot()` section : https://github.com/Laravel-Backpack/Settings/blob/master/src/SettingsServiceProvider.php#L82

Laravel documentation mentions this:

> You should never attempt to register any event listeners, routes, or any other piece of functionality within the register method. Otherwise, you may accidentally use a service that is provided by a service provider which has not loaded yet.

https://laravel.com/docs/5.5/providers

**Comment by pavoltanuska** [comment]
@tabacitu the order seems to be correct (it was the first thing I've checked)

```
"require": {
        "php": ">=7.0.0",
        "backpack/base": "^0.8.0",
        "backpack/crud": "^3.3.0",
        "backpack/settings": "^2.0",
        ...
}
```

The weird thing is, I've tried to revert my changes back to the default state and now everything works without any issue.

Despite that, @jguyomard's suggestion seems to be correct and should solve this behaviour in the end, so we can probably close this issue after merging his PR.

Thanks!
